src/Http/Http/src/Features/ResponseBodyPipeFeature.cs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.IO.Pipelines;
+
+namespace Microsoft.AspNetCore.Http.Features
+{
+    public class ResponseBodyPipeFeature : IResponseBodyPipeFeature
+    {
+        private readonly static Func<IFeatureCollection, IHttpResponseFeature> _nullRequestFeature = f => null;
+
+        private PipeWriter _pipeWriter;
+        private FeatureReferences<IHttpResponseFeature> _features;
+
+        public ResponseBodyPipeFeature(IFeatureCollection features)
+        {
+            if (features == null)
+            {
+                throw new ArgumentNullException(nameof(features));
+            }
+
+            _features = new FeatureReferences<IHttpResponseFeature>(features);
+        }
+
+        private IHttpResponseFeature HttpResponseFeature =>
+            _features.Fetch(ref _features.Cache, _nullRequestFeature);
+
+        public PipeWriter PipeWriter
+        {
+            get
+            {
+                if (_pipeWriter == null ||
+                    // If the Response.Body has been updated, recreate the pipeWriter
+                    (_pipeWriter is StreamPipeWriter writer && !object.ReferenceEquals(writer.InnerStream, HttpResponseFeature.Body)))
+                {
+                    _pipeWriter = new StreamPipeWriter(HttpResponseFeature.Body);
```

Comment:
In this case, do we lose the old reference to the pipe? Who is responsible for cleaning it up? Can this leak memory?

Reply:
Tough situation. Someone can keep a reference to the BodyPipe like:

```c#
  context.Response.Body = new MemoryStream();
  var canBeDisposed = context.Response.BodyPipe;
  context.Response.Body = Stream.Null;
  var nonDisposed = context.Response.BodyPipe;
```

Should we dispose canBeDisposed in this situation? I don't think we would dispose the stream in this situation, so I don't think we should call dispose here.

Reply:
Considering we're implicitly constructing the StreamPipeWriter, we should dispose it. I don't think we can expect developers to know they might have dispose the old response body StreamPipeWriter just because they replaced the response body Stream.

In this case it's not trivial to track who constructed the _pipeWriter, but we should try. It's also probably a good idea to fail any I/O being done on the old _pipeWriter instead of closing in gracefully.

Reply:
>In this case it's not trivial to track who constructed the _pipeWriter, but we should try.

I think the tracking needs to exist inside of the feature. I'll take a stab at something.
> It's also probably a good idea to fail any I/O being done on the old _pipeWriter instead of closing in gracefully.

You are effectively asking for us to call Abort on the underlying stream, correct?

Reply:
Also, what happens if someone wraps the BodyPipe with another PipeReader/Writer. For example, if we were to reimplement ResponseCompression with Pipes, we would do:
```c#
ctx.Response.BodyPipe = new ResponseCompressionPipeWriter(ctx.Response.BodyPipe);
```

which would fail because the BodyPipe is now disposed.